File: backend/app/schemas/growth_observation.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class StressSourceRead(BaseModel):
    code: str
    label: str
    evidence: str = ""
    count: str = "1"


class EmotionTrendRead(BaseModel):
    direction: str = "stable"
    label: str = "稳定"
    signals: list[str] = []


class TeacherGuidanceRead(BaseModel):
    need_attention: bool = False
    urgency: str = "observe"
    urgency_label: str = "建议观察"
    suggested_actions: list[str] = []
    duration_assessment: str = ""
    rationale: str = ""


class AnalysisWindowRead(BaseModel):
    date_from: str = Field(alias="from")
    date_to: str = Field(alias="to")
    days: int = 7
    moment_count: int = 0
    report_days: int = 0

    model_config = {"populate_by_name": True}


class GrowthObservationReportRead(BaseModel):
    risk_tier: str = "none"
    risk_tier_label: str = "无风险"
    risk_summary: list[str] = []
    stress_sources: list[StressSourceRead] = []
    emotion_trend: EmotionTrendRead = Field(default_factory=EmotionTrendRead)
    teacher_guidance: TeacherGuidanceRead = Field(default_factory=TeacherGuidanceRead)
    student_weekly_hint: str = ""
    disclaimer: str = ""
    analysis_window: AnalysisWindowRead | None = None


class StudentGrowthObservationRead(BaseModel):
    weekly_hint: str = ""
    trend_label: str = "稳定"
    stress_directions: list[str] = []
    disclaimer: str = ""


class GrowthInsightRead(BaseModel):
    status: str
    focus_tags: list[str] = []
    focus_directions: list[str] = []
    trend: str = "stable"
    summary: str = ""
    need_attention: bool = False
    risk_level: str = "none"
    risk_reminder: str | None = None


class AttentionTagRead(BaseModel):
    code: str
    label: str
    count: int = 1


class GrowthFocusRead(BaseModel):
    id: uuid.UUID
    student_id: uuid.UUID
    student_name: str | None = None
    class_name: str | None = None
    report_date: str | None = None
    date_end: str
    title: str
    growth_status: str
    summary: str
    focus_directions: list[str] = []
    focus_tags: list[str] = []
    trend: str = "stable"
    need_attention: bool = False
    risk_level: str = "none"
    risk_reminder: str | None = None
    follow_up_status: str
    acked_at: datetime | None = None
    alert_type: str


class TrendPointRead(BaseModel):
    date: str
    mood_score: float
    record_count: int


class GrowthTimelineItemRead(BaseModel):
    moment_id: uuid.UUID | None = None
    date: str
    emotion_tag: str
    category_tag: str | None = None
    category_label: str | None = None
    story_detail: str | None = None
    note: str | None = None
    note_exposed: bool = False
    can_dismiss: bool = False
    ai_tags: list[str] = []


class DangerSignalRecordRead(BaseModel):
    moment_id: uuid.UUID
    date: str
    category_tag: str | None = None
    category_label: str
    story_detail: str
    note: str
    emotion_tag: str
    can_dismiss: bool = True


class CriticalRiskFollowCreate(BaseModel):
    note: str | None = Field(default=None, max_length=2000)


class CriticalRiskFollowStateRead(BaseModel):
    moment_id: uuid.UUID
    follow_up_status: str
    follow_note: str | None = None
    followed_at: datetime | None = None


class CriticalRiskSignalRead(BaseModel):
    moment_id: uuid.UUID
    student_id: uuid.UUID
    student_name: str
    class_name: str
    report_date: str
    category_label: str
    story_detail: str
    note_preview: str
    emotion_tag: str
    risk_reminder: str
    follow_up_status: str = "pending"
    follow_note: str | None = None
    followed_at: datetime | None = None


class CriticalRiskDetailRead(BaseModel):
    moment_id: uuid.UUID
    student_id: uuid.UUID
    student_name: str
    class_name: str
    report_date: str
    emotion_tag: str
    category_label: str
    detail_tags: list[str] = []
    story_detail: str
    note: str
    companion_scene: str
    companion_scene_label: str = ""
    created_at: datetime
    can_dismiss: bool = True
    risk_reminder: str
    follow_up_status: str = "pending"
    follow_note: str | None = None
    followed_at: datetime | None = None


class RiskExposureRead(BaseModel):
    moment_id: uuid.UUID
    date: str
    emotion_tag: str
    note: str
    can_dismiss: bool = True


class RiskDismissRead(BaseModel):
    student_id: uuid.UUID
    moment_id: uuid.UUID
    report_date: str
    risk_level: str
    risk_reminder: str | None = None


class TeacherFollowUpRead(BaseModel):
    id: uuid.UUID
    action: str
    note: str | None = None
    created_at: datetime


class TeacherFollowUpCreate(BaseModel):
    action: str = Field(pattern="^(communicated|contacted_parent|referred_counselor|observing)$")
    note: str | None = Field(default=None, max_length=2000)


class GrowthDayRecordRead(BaseModel):
    date: str
    ai_summary: str
    insight: GrowthInsightRead
    moment_count: int = 0
    mood_counts: dict[str, int] = {}
    category_breakdown: dict[str, int] = {}
    mood_score: float | None = None
    has_report: bool = False
    danger_records: list[DangerSignalRecordRead] = []
    entries: list[GrowthTimelineItemRead] = []


class GrowthArchiveRead(BaseModel):
    student_id: uuid.UUID
    student_name: str
    class_name: str
    ai_summary: str
    insight: GrowthInsightRead
    observation: GrowthObservationReportRead | None = None
    trend_metric_label: str = ""
    trend_points: list[TrendPointRead]
    mood_counts: dict[str, int]
    category_breakdown: dict[str, int]
    mood_counts_by_category: dict[str, dict[str, int]] = {}
    attention_tags: list[AttentionTagRead]
    daily_records: list[GrowthDayRecordRead] = []
    timeline: list[GrowthTimelineItemRead] = []
    risk_exposures: list[RiskExposureRead] = []
    follow_ups: list[TeacherFollowUpRead]

File: teacher_app/lib/data/models/growth_observation.dart
import 'package:flutter/material.dart';

import 'critical_risk.dart';

class StressSource {
  StressSource({
    required this.code,
    required this.label,
    this.evidence = '',
    this.count = '1',
  });

  final String code;
  final String label;
  final String evidence;
  final String count;

  factory StressSource.fromJson(Map<String, dynamic> json) => StressSource(
        code: json['code'] as String? ?? '',
        label: json['label'] as String? ?? '',
        evidence: json['evidence'] as String? ?? '',
        count: json['count'] as String? ?? '1',
      );
}

class EmotionTrendDetail {
  EmotionTrendDetail({
    required this.direction,
    required this.label,
    this.signals = const [],
  });

  final String direction;
  final String label;
  final List<String> signals;

  factory EmotionTrendDetail.fromJson(Map<String, dynamic> json) => EmotionTrendDetail(
        direction: json['direction'] as String? ?? 'stable',
        label: json['label'] as String? ?? '稳定',
        signals: (json['signals'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
      );
}

class TeacherGuidance {
  TeacherGuidance({
    required this.needAttention,
    required this.urgency,
    required this.urgencyLabel,
    this.suggestedActions = const [],
    this.durationAssessment = '',
    this.rationale = '',
  });

  final bool needAttention;
  final String urgency;
  final String urgencyLabel;
  final List<String> suggestedActions;
  final String durationAssessment;
  final String rationale;

  factory TeacherGuidance.fromJson(Map<String, dynamic> json) => TeacherGuidance(
        needAttention: json['need_attention'] as bool? ?? false,
        urgency: json['urgency'] as String? ?? 'observe',
        urgencyLabel: json['urgency_label'] as String? ?? '建议观察',
        suggestedActions:
            (json['suggested_actions'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
        durationAssessment: json['duration_assessment'] as String? ?? '',
        rationale: json['rationale'] as String? ?? '',
      );
}

class GrowthObservationReport {
  GrowthObservationReport({
    required this.riskTier,
    required this.riskTierLabel,
    required this.riskSummary,
    required this.stressSources,
    required this.emotionTrend,
    required this.teacherGuidance,
    this.studentWeeklyHint = '',
    this.disclaimer = '',
  });

  final String riskTier;
  final String riskTierLabel;
  final List<String> riskSummary;
  final List<StressSource> stressSources;
  final EmotionTrendDetail emotionTrend;
  final TeacherGuidance teacherGuidance;
  final String studentWeeklyHint;
  final String disclaimer;

  factory GrowthObservationReport.fromJson(Map<String, dynamic> json) =>
      GrowthObservationReport(
        riskTier: json['risk_tier'] as String? ?? 'none',
        riskTierLabel: json['risk_tier_label'] as String? ?? '无风险',
        riskSummary: (json['risk_summary'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
        stressSources: (json['stress_sources'] as List<dynamic>? ?? [])
            .map((e) => StressSource.fromJson(e as Map<String, dynamic>))
            .toList(),
        emotionTrend:
            EmotionTrendDetail.fromJson(json['emotion_trend'] as Map<String, dynamic>? ?? {}),
        teacherGuidance:
            TeacherGuidance.fromJson(json['teacher_guidance'] as Map<String, dynamic>? ?? {}),
        studentWeeklyHint: json['student_weekly_hint'] as String? ?? '',
        disclaimer: json['disclaimer'] as String? ?? '',
      );
}

class GrowthInsight {
  GrowthInsight({
    required this.status,
    required this.focusTags,
    required this.focusDirections,
    required this.trend,
    required this.summary,
    required this.needAttention,
    required this.riskLevel,
    this.riskReminder,
  });

  final String status;
  final List<String> focusTags;
  final List<String> focusDirections;
  final String trend;
  final String summary;
  final bool needAttention;
  final String riskLevel;
  final String? riskReminder;

  factory GrowthInsight.fromJson(Map<String, dynamic> json) => GrowthInsight(
        status: json['status'] as String? ?? 'observing',
        focusTags: (json['focus_tags'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
        focusDirections:
            (json['focus_directions'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
        trend: json['trend'] as String? ?? 'stable',
        summary: json['summary'] as String? ?? '',
        needAttention: json['need_attention'] as bool? ?? false,
        riskLevel: json['risk_level'] as String? ?? 'none',
        riskReminder: json['risk_reminder'] as String?,
      );
}

class GrowthFocusItem {
  GrowthFocusItem({
    required this.id,
    required this.studentId,
    required this.studentName,
    required this.className,
    required this.reportDate,
    required this.dateEnd,
    required this.title,
    required this.growthStatus,
    required this.summary,
    required this.focusDirections,
    required this.focusTags,
    required this.trend,
    required this.needAttention,
    required this.riskLevel,
    required this.followUpStatus,
    this.riskReminder,
    this.ackedAt,
  });

  final String id;
  final String studentId;
  final String? studentName;
  final String? className;
  final String? reportDate;
  final String dateEnd;
  final String title;
  final String growthStatus;
  final String summary;
  final List<String> focusDirections;
  final List<String> focusTags;
  final String trend;
  final bool needAttention;
  final String riskLevel;
  final String? riskReminder;
  final String followUpStatus;
  final DateTime? ackedAt;

  bool get isFollowed => followUpStatus == 'followed';
  bool get isCritical => riskLevel == 'critical';

  String get moodLookupDate => reportDate ?? dateEnd;

  factory GrowthFocusItem.fromJson(Map<String, dynamic> json) {
    DateTime? acked;
    final raw = json['acked_at'];
    if (raw is String) acked = DateTime.tryParse(raw);
    return GrowthFocusItem(
      id: json['id'] as String? ?? '',
      studentId: json['student_id'] as String? ?? '',
      studentName: json['student_name'] as String?,
      className: json['class_name'] as String?,
      reportDate: json['report_date'] as String?,
      dateEnd: json['date_end'] as String? ?? '',
      title: json['title'] as String? ?? '成长关注',
      growthStatus: json['growth_status'] as String? ?? 'ongoing',
      summary: json['summary'] as String? ?? '',
      focusDirections:
          (json['focus_directions'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
      focusTags: (json['focus_tags'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
      trend: json['trend'] as String? ?? 'stable',
      needAttention: json['need_attention'] as bool? ?? false,
      riskLevel: json['risk_level'] as String? ?? 'none',
      riskReminder: json['risk_reminder'] as String?,
      followUpStatus: json['follow_up_status'] as String? ?? 'pending',
      ackedAt: acked,
    );
  }
}

class AttentionTag {
  AttentionTag({required this.code, required this.label, required this.count});
  final String code;
  final String label;
  final int count;

  factory AttentionTag.fromJson(Map<String, dynamic> json) => AttentionTag(
        code: json['code'] as String? ?? '',
        label: json['label'] as String? ?? '',
        count: json['count'] as int? ?? 1,
      );
}

class TrendPoint {
  TrendPoint({required this.date, required this.moodScore, required this.recordCount});
  final String date;
  final double moodScore;
  final int recordCount;

  factory TrendPoint.fromJson(Map<String, dynamic> json) => TrendPoint(
        date: json['date'] as String? ?? '',
        moodScore: (json['mood_score'] as num?)?.toDouble() ?? 0,
        recordCount: json['record_count'] as int? ?? 0,
      );
}

class GrowthTimelineItem {
  GrowthTimelineItem({
    this.momentId,
    required this.date,
    required this.emotionTag,
    this.categoryTag,
    this.categoryLabel,
    this.note,
    this.noteExposed = false,
    this.canDismiss = false,
    required this.aiTags,
  });

  final String? momentId;
  final String date;
  final String emotionTag;
  final String? categoryTag;
  final String? categoryLabel;
  final String? note;
  final bool noteExposed;
  final bool canDismiss;
  final List<String> aiTags;

  factory GrowthTimelineItem.fromJson(Map<String, dynamic> json) => GrowthTimelineItem(
        momentId: json['moment_id']?.toString(),
        date: json['date'] as String? ?? '',
        emotionTag: json['emotion_tag'] as String? ?? '',
        categoryTag: json['category_tag'] as String?,
        categoryLabel: json['category_label'] as String?,
        note: json['note'] as String?,
        noteExposed: json['note_exposed'] as bool? ?? false,
        canDismiss: json['can_dismiss'] as bool? ?? false,
        aiTags: (json['ai_tags'] as List<dynamic>? ?? []).map((e) => '$e').toList(),
      );
}

class RiskExposure {
  RiskExposure({
    required this.momentId,
    required this.date,
    required this.emotionTag,
    required this.note,
    this.canDismiss = true,
  });

  final String momentId;
  final String date;
  final String emotionTag;
  final String note;
  final bool canDismiss;

  factory RiskExposure.fromJson(Map<String, dynamic> json) => RiskExposure(
        momentId: json['moment_id']?.toString() ?? '',
        date: json['date'] as String? ?? '',
        emotionTag: json['emotion_tag'] as String? ?? '',
        note: json['note'] as String? ?? '',
        canDismiss: json['can_dismiss'] as bool? ?? true,
      );
}

class TeacherFollowUp {
  TeacherFollowUp({required this.id, required this.action, this.note, required this.createdAt});
  final String id;
  final String action;
  final String? note;
  final DateTime createdAt;

  factory TeacherFollowUp.fromJson(Map<String, dynamic> json) {
    final raw = json['created_at'];
    return TeacherFollowUp(
      id: json['id'] as String? ?? '',
      action: json['action'] as String? ?? '',
      note: json['note'] as String?,
      createdAt: raw is String ? DateTime.parse(raw) : DateTime.now(),
    );
  }
}

class GrowthDayRecord {
  GrowthDayRecord({
    required this.date,
    required this.aiSummary,
    required this.insight,
    required this.momentCount,
    required this.moodCounts,
    required this.categoryBreakdown,
    this.moodScore,
    this.hasReport = false,
    this.dangerRecords = const [],
    this.entries = const [],
  });

  final String date;
  final String aiSummary;
  final GrowthInsight insight;
  final int momentCount;
  final Map<String, int> moodCounts;
  final Map<String, int> categoryBreakdown;
  final double? moodScore;
  final bool hasReport;
  final List<DangerSignalRecord> dangerRecords;
  final List<GrowthTimelineItem> entries;

  factory GrowthDayRecord.fromJson(Map<String, dynamic> json) => GrowthDayRecord(
        date: json['date'] as String? ?? '',
        aiSummary: json['ai_summary'] as String? ?? '',
        insight: GrowthInsight.fromJson(json['insight'] as Map<String, dynamic>? ?? {}),
        momentCount: json['moment_count'] as int? ?? 0,
        moodCounts: (json['mood_counts'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, (v as num).toInt())),
        categoryBreakdown: (json['category_breakdown'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, (v as num).toInt())),
        moodScore: (json['mood_score'] as num?)?.toDouble(),
        hasReport: json['has_report'] as bool? ?? false,
        dangerRecords: (json['danger_records'] as List<dynamic>? ??
                json['entries'] as List<dynamic>? ??
                [])
            .map((e) => DangerSignalRecord.fromJson(e as Map<String, dynamic>))
            .toList(),
        entries: (json['entries'] as List<dynamic>? ?? [])
            .map((e) => GrowthTimelineItem.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

class GrowthArchive {
  GrowthArchive({
    required this.studentId,
    required this.studentName,
    required this.className,
    required this.aiSummary,
    required this.insight,
    this.observation,
    this.trendMetricLabel = '',
    required this.trendPoints,
    required this.moodCounts,
    required this.categoryBreakdown,
    this.moodCountsByCategory = const {},
    required this.attentionTags,
    this.dailyRecords = const [],
    this.timeline = const [],
    this.riskExposures = const [],
    this.followUps = const [],
  });

  final String studentId;
  final String studentName;
  final String className;
  final String aiSummary;
  final GrowthInsight insight;
  final GrowthObservationReport? observation;
  final String trendMetricLabel;
  final List<TrendPoint> trendPoints;
  final Map<String, int> moodCounts;
  final Map<String, int> categoryBreakdown;
  final Map<String, Map<String, int>> moodCountsByCategory;
  final List<AttentionTag> attentionTags;
  final List<GrowthDayRecord> dailyRecords;
  final List<GrowthTimelineItem> timeline;
  final List<RiskExposure> riskExposures;
  final List<TeacherFollowUp> followUps;

  factory GrowthArchive.fromJson(Map<String, dynamic> json) => GrowthArchive(
        studentId: json['student_id'] as String? ?? '',
        studentName: json['student_name'] as String? ?? '',
        className: json['class_name'] as String? ?? '',
        aiSummary: json['ai_summary'] as String? ?? '',
        insight: GrowthInsight.fromJson(json['insight'] as Map<String, dynamic>? ?? {}),
        observation: json['observation'] is Map<String, dynamic>
            ? GrowthObservationReport.fromJson(json['observation'] as Map<String, dynamic>)
            : null,
        trendMetricLabel: json['trend_metric_label'] as String? ?? '',
        trendPoints: (json['trend_points'] as List<dynamic>? ?? [])
            .map((e) => TrendPoint.fromJson(e as Map<String, dynamic>))
            .toList(),
        moodCounts: (json['mood_counts'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, (v as num).toInt())),
        categoryBreakdown: (json['category_breakdown'] as Map<String, dynamic>? ?? {})
            .map((k, v) => MapEntry(k, (v as num).toInt())),
        moodCountsByCategory: _parseMoodCountsByCategory(json['mood_counts_by_category']),
        attentionTags: (json['attention_tags'] as List<dynamic>? ?? [])
            .map((e) => AttentionTag.fromJson(e as Map<String, dynamic>))
            .toList(),
        dailyRecords: (json['daily_records'] as List<dynamic>? ?? [])
            .map((e) => GrowthDayRecord.fromJson(e as Map<String, dynamic>))
            .toList(),
        timeline: (json['timeline'] as List<dynamic>? ?? [])
            .map((e) => GrowthTimelineItem.fromJson(e as Map<String, dynamic>))
            .toList(),
        riskExposures: (json['risk_exposures'] as List<dynamic>? ?? [])
            .map((e) => RiskExposure.fromJson(e as Map<String, dynamic>))
            .toList(),
        followUps: (json['follow_ups'] as List<dynamic>? ?? [])
            .map((e) => TeacherFollowUp.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

Map<String, Map<String, int>> _parseMoodCountsByCategory(dynamic raw) {
  if (raw is! Map) return {};
  return raw.map((catKey, catVal) {
    if (catVal is! Map) return MapEntry('$catKey', <String, int>{});
    return MapEntry(
      '$catKey',
      catVal.map((k, v) => MapEntry('$k', (v as num).toInt())),
    );
  });
}

String growthStatusLabel(String status) {
  switch (status) {
    case 'priority':
      return '优先关注';
    case 'ongoing':
      return '持续关注';
    default:
      return '观察中';
  }
}

String followUpActionLabel(String action) {
  switch (action) {
    case 'communicated':
      return '已沟通';
    case 'contacted_parent':
      return '已联系家长';
    case 'referred_counselor':
      return '已转心理老师';
    default:
      return '持续观察';
  }
}

String trendLabel(String trend) {
  switch (trend) {
    case 'up':
      return '上升';
    case 'down':
      return '下降';
    case 'worsening':
      return '逐渐变差';
    case 'significantly_worsening':
      return '明显恶化';
    default:
      return '稳定';
  }
}

Color riskTierColor(String tier) {
  switch (tier) {
    case 'urgent':
      return const Color(0xFFD32F2F);
    case 'high':
      return const Color(0xFFE65100);
    case 'moderate':
      return const Color(0xFFFF9800);
    case 'light':
      return const Color(0xFFFFB74D);
    default:
      return const Color(0xFF7CB342);
  }
}

String riskTierLabel(String tier) {
  switch (tier) {
    case 'urgent':
      return '紧急关注';
    case 'high':
      return '高度关注';
    case 'moderate':
      return '中度关注';
    case 'light':
      return '轻度关注';
    default:
      return '无风险';
  }
}
